Core/Src/ACDC_SPI.c

```diff
@@ -0,0 +1,96 @@
+/**
+ * @file ACDC_SPI.h
+ * @author Liam Bramley
+ * @brief Implementation of SPI driver
+ * 
+ * This file includes functions to initalize the SPI, transfer and recieve data.
+ * 
+ * @version 0.1
+ * @date 2024-01-30
+ * @copyright Copyright (c) 2023-2024
+ * 
+ */
+
+#include "ACDC_SPI.h"
+
+
+/*
+{
+  __IO uint32_t CR1;      /* SPI control register 1         */
+  __IO uint32_t CR2;      /* SPI control register 2         */
+  __IO uint32_t SR;       /* SPI status register            */
+  __IO uint32_t DR;       /* SPI data register              */
+  __IO uint32_t CRCPR;    /* SPI CRC polynomial register    */
+  __IO uint32_t RXCRCR;   /* SPI RX CRC register            */
+  __IO uint32_t TXCRCR;   /* SPI TX CRC register            */
+  __IO uint32_t I2SCFGR;  /* SPI_I2S configuration register 
+ SPI_TypeDef;
+*/
+
+void SPI_Initalize(const SPI_TypeDef *SPI) {
+    if(SPI == SPI1){
+        // Enable clock for SPI
+        RCC->APB2ENR |= RCC_APB2ENR_SPI1EN;
+
+        // Configure GPIO pins for SPI1
+        // Set PA5 (SCK) as alternate function push-pull output
+        GPIO_PinDirection(GPIOA, GPIO_PIN_5, GPIO_MODE_OUTPUT_SPEED_50MHz, GPIO_CNF_OUTPUT_AF_PUSH_PULL);
+
+        // Set PA6 (MISO) as input floating
+        GPIO_PinDirection(GPIOA, GPIO_PIN_6, GPIO_MODE_INPUT, GPIO_CNF_INPUT_FLOATING);
+
+        // Set PA7 (MOSI) as alternate function push-pull output
+        GPIO_PinDirection(GPIOA, GPIO_PIN_7, GPIO_MODE_OUTPUT_SPEED_50MHz, GPIO_CNF_OUTPUT_AF_PUSH_PULL);
+
+        // Configure SPI1
+        SPI1->CR1 = 0; // Clear CR1 register
+        SPI1->CR1 |= SPI_CR1_SSM | SPI_CR1_SSI | SPI_CR1_BR_0 | SPI_CR1_BR_1; // Set master mode, software NSS, baud rate
```

Comment:
It also looks like the Data Frame Format (DFF) was left to its default value (8-bits). We may need a 16-bit format so that we have to write/run less code. 
[{See RM-742}](https://www.st.com/resource/en/reference_manual/rm0008-stm32f101xx-stm32f102xx-stm32f103xx-stm32f105xx-and-stm32f107xx-advanced-armbased-32bit-mcus-stmicroelectronics.pdf#page=742)

Here is some example code you can use

## ACDC_SPI.h

```c
typedef enum{   // SPI Data Frame Format
    SPI_MODE_8Bit  = 0,         /**< 8-bit data frame format for Tx/Rx  */
    SPI_MODE_16Bit = 1          /**< 16-bit data frame format for Tx/Rx */
}SPI_BitMode;

/// @brief Changes the SPIx peripheral between 8-bit and 16-bit mode
/// @param SPIx SPI Peripheral (Ex. SPI1 or SPI2)
/// @param SPI_MODE_x Bit mode selection for SPIx (Ex. SPI_MODE_8bit or SPI_MODE_16bit)
void SPI_SetBitMode(SPI_TypeDef *SPIx, SPI_BitMode SPI_MODE_x);
``` 

## ACDC_SPI.c

```c
void SPI_SetBitMode(SPI_TypeDef *SPIx, SPI_BitMode SPI_MODE_x){
    CLEAR_BIT(SPIx->CR1, SPI_CR1_SPE);  // Disable SPIx
    
    if(SPI_MODE_x == SPI_MODE_8Bit)         // If the user wants 8-bit mode
        CLEAR_BIT(SPIx->CR1, SPI_CR1_DFF);  // Set to 8-bit mode
    else if(SPI_MODE_x == SPI_MODE_16Bit)   // If the user wants 16-bit mode
        SET_BIT(SPIx->CR1, SPI_CR1_DFF);    // Set to 16-bit mode

    SET_BIT(SPIx->CR1, SPI_CR1_SPE);    // Enable SPIx
}
``` 

> **_NOTE:_** The DFF bit should only be written only when SPIx is disabled (SPE = ‘0’) for correct operation.